### ISSUE TITLE
[perf] SearchExpressionFacade should not use VisitTree

### ISSUE THREAD

**Issue by martin654** [primefaces/primefaces#1827]
## 1) Environment
- PrimeFaces version: 6.0 
- Does it work on the newest PrimeFaces version? no
- Application server + version: Tomcat 8.0 (also tested on WAS 8.5.5 and it is the same)
- Affected browsers: All
## 2) Expected behavior

Page is loaded with Primefaces 6.0 at the same speed as in 5.3
## 3) Actual behavior

Page loading takes much more longer. In given example it takes approximately 3-5 seconds with Primefaces 5.3 and about 20 seconds with Primefaces 6.0.
## 4) Steps to reproduce
- Create composite componens as described below in Sample XHTML. 
- Create XHTML page with a lot of those composite components (2000 in this example) which are displayed using for example p:dataGrid. 
- Visit that page.
- Observe difference in performance between PF 5.3 and PF 6.0.
## 5) Sample XHTML page

Page with repetition using p:dataGrid:

```
<h:form id="form">

      <div class="Container ContainerIndent">
        <p:commandButton id="cb_back" ajax="false" action="back" icon="fa fa-arrow-right" value="Back to wizard" 
          iconPos="right" styleClass="BB" immediate="true" />
      </div>

     <div class="content">       
        <p:panel styleClass="ContainerIndent" header="Dynaimc form">

          <p:dataGrid id="parameterDataGrid"  value="#{dynamicFormModel.listV07}" var="dynamicField" 
            columns="5" layout="grid" styleClass="no-border" >

            <dynamicform:testInputText id="inp" value="#{dynamicField}" label="Label for #{dynamicField}"/>

          </p:dataGrid>
        </p:panel>
     </div>
    </h:form>
```

Composite testInputText:

```
<ui:component xmlns="http://www.w3.org/1999/xhtml" xmlns:cc="http://xmlns.jcp.org/jsf/composite"
  xmlns:p="http://primefaces.org/ui" xmlns:ui="http://xmlns.jcp.org/jsf/facelets" 
  xmlns:dynamicform="http://xmlns.jcp.org/jsf/composite/flows/showcase/dynamicform">

  <cc:interface shortDescription="testInputText is an extension to a standard primefces inputText.">
    <cc:attribute name="value" required="true" shortDescription="Required. Field on model for value."/>
    <cc:attribute name="label" required="true" type="java.lang.String" shortDescription="Required. Label to show above component."/>
    <cc:editableValueHolder name="inputText" />
  </cc:interface>

  <cc:implementation>
    <ui:param name="componentId" value="#{cc.clientId}:inputText" />

    <div class="fob-input fob-component" id="#{cc.clientId}">
      <div class="fob-label-row">
        <dynamicform:testLabel id="inputTextLabel" forField="#{componentId}" label="#{cc.attrs.label}"  />
      </div>
      <div class="fob-value-row">
        <p:inputText id="inputText" value="#{cc.attrs.value}" label="#{cc.attrs.label}"
          required="#{cc.attrs.required}" disabled="#{cc.attrs.disabled}"
          styleClass="fob-input" >
          <cc:insertChildren />
        </p:inputText>
      </div>
    </div>
  </cc:implementation>
</ui:component>
```

Composite testLabel:

```
<?xml version="1.0" encoding="UTF-8"?>
<ui:component xmlns="http://www.w3.org/1999/xhtml" xmlns:ui="http://xmlns.jcp.org/jsf/facelets" 
  xmlns:cc="http://xmlns.jcp.org/jsf/composite" xmlns:p="http://primefaces.org/ui" >

  <cc:interface name="label" expert="true"
    shortDescription="Component representing label. Label component consists of tooltip and label. Used in composite components.">
    <cc:attribute name="forField" required="true" 
      shortDescription="Component id for which is the tooltip displayed."/>
    <cc:attribute name="label" required="true" 
      shortDescription="Label to show."/>
  </cc:interface>

  <cc:implementation>
    <!-- Label -->
    <span class="fob-output-label">
      <p:outputLabel id="label" value="#{cc.attrs.label}" for="#{cc.attrs.forField}"/>
    </span>
  </cc:implementation>
</ui:component>
```
## 6) Sample bean

Model (created before page is visited):

```
private List<String> listV07;

public List<String> getListV07() {
  return listV07;
}

public void setListV07(List<String> listV07) {
  this.listV07 = listV07;
}
```

Model is created using this code:

```
public void initModel() {
  model.setListV07(createListV07(2000));
}

private List<String> createListV07(int count) {
  List<String> list = new ArrayList<>();
  for (int i = 0; i < count; i++) {
    list.add("Some text " + i);
  }
  return list;
}
```
## 7) Detailed description

I think, that it has to do something with p:outputLabel and it's for attribute when used in composite components.

I profiled my application using YourKit profiler and I found out, that extreme amount of time with PF 6.0 is spent in javax.faces.component.UIComponent.visitTree(VisitContext, VisitCallback) method.

Interesting part of call tree for PF 6.0 is:
-UIComponentBase.java:920org.primefaces.component.outputlabel.OutputLabelRenderer.encodeEnd(FacesContext, UIComponent)
-OutputLabelRenderer.java:59org.primefaces.expression.SearchExpressionFacade.resolveComponent(FacesContext, UIComponent, String)
-SearchExpressionFacade.java:377org.primefaces.expression.SearchExpressionFacade.resolveComponent(FacesContext, UIComponent, String, int)
-SearchExpressionFacade.java:425org.primefaces.expression.SearchExpressionFacade.resolveComponentById(UIComponent, String, String, FacesContext, int)
-SearchExpressionFacade.java:489org.primefaces.util.ComponentTraversalUtils.firstById(String, UIComponent, String, FacesContext, boolean) 
-ComponentTraversalUtils.java:172javax.faces.component.UIComponent.visitTree(VisitContext, VisitCallback)

Using PF 5.3 the part of call tree is different:
-UIComponentBase.java:920org.primefaces.component.outputlabel.OutputLabelRenderer.encodeEnd(FacesContext, UIComponent)
-OutputLabelRenderer.java:59org.primefaces.expression.SearchExpressionFacade.resolveComponent(FacesContext, UIComponent, String)
-SearchExpressionFacade.java:375org.primefaces.expression.SearchExpressionFacade.resolveComponent(FacesContext, UIComponent, String, int)
-SearchExpressionFacade.java:423org.primefaces.expression.SearchExpressionFacade.resolveComponentById(UIComponent, String, String, FacesContext, int)
-SearchExpressionFacade.java:487org.primefaces.util.ComponentTraversalUtils.firstById(String, UIComponent, String, FacesContext, boolean)
-ComponentTraversalUtils.java:158javax.faces.component.UIComponentBase.invokeOnComponent(FacesContext, String, ContextCallback)

I noticed that in _org.primefaces.util.ComponentTraversalUtils_ class in method: _public static UIComponent firstById(String id, UIComponent base, String separatorString, FacesContext context, boolean skipUnrendered)_
the code evaluates dirrefently the condition **_if (skipUnrendered)_**
In PF 5.3 it evaluates as false and in PF 6.0 it evaluates as true. 

Hint:
I found out that this code was affected by this issue: https://github.com/primefaces/primefaces/issues/1102 where default behaviour affecting condition (skipUnrendered) was changed
in class _org.primefaces.expression.SearchExpressionFacade_ in method _private static UIComponent resolveComponentById(UIComponent source, String expression, String separatorString, FacesContext context, int options)_

_isOptionSet(options, Options.IGNORE_NO_RESULT)_ 
was changed to 
_!isOptionSet(options, Options.VISIT_UNRENDERED)_

**Comment by tandraschko** [comment]
The following happens in trunk:

1) OutputLabelRenderer calls SearchExpressionFacade#resolveComponent without a option
2) #resolveComponentById is called with Options.None
3) !isOptionSet(options, Options.VISIT_UNRENDERED) is evaluated to false
4) ComponentTraversalUtils#firstById is called with skipUnrendered=false

This means that 3) is wrong and should be without "1". Right?

**Comment by martin654** [comment]
1) yes
2) yes
3) isOptionSet(options, Options.VISIT_UNRENDERED) is evaluated to false, and !isOptionSet(options, Options.VISIT_UNRENDERED) is evaluated to **TRUE**
4) ComponentTraversalUtils.firstById is called with skipUnrendered = true (in version 5.3 it was called with skipUnrendered = false) and this causes different behaviour (calling visitTree)

**Comment by tandraschko** [comment]
Hmm, so i don't understand why it would be slower.
skipUnrendered=true should bring the better performance.

if skipUnrendered is set, we pass VisitHint.SKIP_UNRENDERED (which should be faster). Everything else is the same.
Could it be that the call to ComponentTraversalUtils#firstById results in null in your case?

**Comment by martin654** [comment]
Well, that is the problem :) It should be faster but it is not. (I found your comment in the issue https://github.com/primefaces/primefaces/issues/1102 that it should improve performance).

The issue is related with the size of component tree and the fact, that ComponentTraversalUtils.firstById method goes through different path.

I think, that with a big number of components (in this case the problem is related with composite components too) the code in IF branch is not efficient. 

I attached screenshot from profiler, where you can see number of invocations of visit tree method.
![yourkit](https://cloud.githubusercontent.com/assets/22746146/19309171/951f69fe-9083-11e6-8040-fe16e6774af3.png)

**Comment by tandraschko** [comment]
Could you please do me a favor and provide 4 screenshots from your results:
1) newest mojarra + skipUnredered=true
2) newest mojarra + skipUnrendered=false
3) newest MyFaces + skipUnredered=true
4) newest MyFaces + skipUnrendered=false

all tests with ProjectStage=Production

The problem could be that isRendered evaluation is not cheap all the time as EL needs to be evaluated sometimes.
In this case, we should pass skipUnrendered=false if ProjectStage==Production.
As skipUnrendered is a good thing to tell people that "you can't refer a unrendered component in this case" and we should not remove it completely.

**Comment by martin654** [comment]
I do those tests in our production app on a test page. I am afraid that testing MyFaces won't be as easy as just updating POM, but I will try.

ProjectStage is (and was) set to production.

The screenshot above is a **case 1)**. Mojarra version 2.2.12 and Primefaces 6.0 (skipUnrendered evaluated to true)

**case 2)**. Mojarra version 2.2.12 and Primefaces 5.3 (skipUnrendered evaluated to false):

![yourkit_mojarra_pf53](https://cloud.githubusercontent.com/assets/22746146/19309984/37d1ebb4-9088-11e6-8867-7e835e138760.png)

**Comment by tandraschko** [comment]
Yep i see. There's a huge different in the visitTree and invokeOnComponent API.
Will rename the VISIT_UNRENDERED a little bit and always use invokeOnComponent.

**Comment by martin654** [comment]
That is great news, that it will be fixed 👍  

Will it be fixed in milestone 6.1 or sooner in a minor PF version?

Thak you for your time. 
Martin

**Comment by tandraschko** [comment]
Not sure, we can ping @mertsincan about that :)
As the behavior is a little bit different now, i personally would not push it into a bugfix version.

**Comment by martin654** [comment]
@mertsincan could you include this fix for 6.0.7 please?

**Comment by martin654** [comment]
@mertsincan ping ;-)

**Comment by martin654** [comment]
@mertsincan could you include this fix for 6.0.8 please?

**Comment by mertsincan** [comment]
> As the behavior is a little bit different now, i personally would not push it into a bugfix version.

I agree with Thomas. Unfortunately, we can not add this fix into elite release.
